### ISSUE TITLE
Allow Prefrence to Rewrite URL Before Request

### ISSUE THREAD

**Issue by kevincox** [requirejs/requirejs#663]
Currently all of the options to set URL's are static.  It would be nice to have the option to set a function that would return the final url to be requested.

An example use case is if you have multiple hosts to serve js files (or just one with multiple host names).  You could use this to get around the browsers default requests per server limit.

```
var i;
function myRewrite(p)
{
    i = (i+1)%4;
    return "//cdn"+i+".example.com/scripts/"+p;
}
```

This would serve request from 4 different host and allow the browser to download more in parallel.

I don't know how you would want to implement it but if could allow `baseUrl` to be a function, and if it is, call it to get the full URL rather than prepending it.
